app/models/pipeline_run.rb

```diff
@@ -153,38 +157,255 @@ def retry
   end
 
   def report_ready?
-    job_status == STATUS_CHECKED || (ready_step && active_stage && active_stage.step_number > ready_step)
+    clause_for_old_runs = (job_status == STATUS_CHECKED || (ready_step && pipeline_run_stages.find_by(step_number: ready_step) && pipeline_run_stages.find_by(step_number: ready_step).job_status == STATUS_LOADED))
+    # TODO: migrate old runs so we don't need to deal with them separately in the code
+    results_finalized == 1 || result_status_for(REPORT_READY_LOADER) || clause_for_old_runs
```

Comment:
why check results_finalized == 1 here?

Reply:
If EVERYTHING has finished, we don't need to bother with the other checks